src/LogWindow.cpp

```diff
@@ -0,0 +1,229 @@
+
+#include "AudacityLogger.h"
+#include "widgets/AudacityMessageBox.h"
+#include "FileNames.h"
+#include "Internat.h"
+#include "MemoryX.h"
+#include "Prefs.h"
+#include "ShuttleGui.h"
+
+#include "../images/AudacityLogoAlpha.xpm"
+
+// If wxLogWindow is used and initialized before the Mac's "root" window, then
+//     Audacity may crash when terminating.  It's not fully understood why this occurs
+//     but it probably has to do with the order of deletion.  However, deferring the
+//     creation of the log window until it is actually shown circumvents the problem.
+enum
+{
+   LoggerID_Save = wxID_HIGHEST + 1,
+   LoggerID_Clear,
+   LoggerID_Close
+};
+
+namespace {
+Destroy_ptr<wxFrame> sFrame;
+wxWeakRef<wxTextCtrl> sText;
+
+struct LogWindowListener : public PrefsListener
+{
+   // PrefsListener implementation
+   void UpdatePrefs() override;
+};
+// Unique PrefsListener can't be statically constructed before the application
+// object initializes, so use Optional
+Optional<LogWindowListener> pListener;
+
+void OnCloseWindow(wxCloseEvent & WXUNUSED(e))
+{
+#if defined(__WXMAC__)
+   // On the Mac, destroy the window rather than hiding it since the
+   // log menu will override the root windows menu if there is no
+   // project window open.
+   sFrame.reset();
+#else
+   LogWindow::Show(false);
+#endif
+}
+
+void OnClose(wxCommandEvent & WXUNUSED(e))
+{
+   wxCloseEvent dummy;
+   OnCloseWindow(dummy);
+}
+
+void OnClear(wxCommandEvent & WXUNUSED(e))
+{
+   auto pLogger = AudacityLogger::Get();
+   if (pLogger)
+      pLogger->ClearLog();
+}
+
+void OnSave(wxCommandEvent & WXUNUSED(e))
+{
+   wxString fName = _("log.txt");
+
+   fName = FileNames::SelectFile(FileNames::Operation::Export,
+      XO("Save log to:"),
+      wxEmptyString,
+      fName,
+      wxT("txt"),
+      { FileNames::TextFiles },
+      wxFD_SAVE | wxFD_OVERWRITE_PROMPT | wxRESIZE_BORDER,
+      sFrame.get());
+
+   if (fName.empty()) {
+      return;
+   }
+
+   if (!(sText && sText->SaveFile(fName))) {
+      AudacityMessageBox(
+         XO("Couldn't save log to file: %s").Format( fName ),
+         XO("Warning"),
+         wxICON_EXCLAMATION,
+         sFrame.get());
+      return;
+   }
+}
+
+void LogWindowListener::UpdatePrefs()
+{
+   //! Re-create the non-modal window in case of change of preferred language
+   if (sFrame) {
+      bool shown = sFrame->IsShown();
+      if (shown) {
+         LogWindow::Show(false);
+      }
+      sFrame.reset();
+      if (shown) {
+         LogWindow::Show(true);
+      }
+   }
+}
+}
+
+void LogWindow::Show(bool show)
+{
+   // Hide the frame if created, otherwise do nothing
+   if (!show) {
+      if (sFrame) {
+         sFrame->Show(false);
+      }
+      return;
+   }
+
+   // If the frame already exists, refresh its contents and show it
+   auto pLogger = AudacityLogger::Get();
+   if (sFrame) {
+      if (!sFrame->IsShown() && sText) {
+         if (pLogger)
+            sText->ChangeValue(pLogger->GetBuffer());
```

Comment:
This method `ChangeValue` still not clearly for me, bind buffer to `sText` object?